pybind_interface/pybind_main.cpp

```diff
@@ -73,264 +82,269 @@ std::vector<Bitstring> getBitstrings(const py::dict &options, int num_qubits) {
                        Circuit<Cirq::GateCirq<float>>* circuit) {
   MakeControlledGate(qubits, values, circuit->gates.back());
 }
 
+template <typename Gate>
+Channel<Gate> create_single_gate_channel(Gate gate) {
+  auto gate_kind = KrausOperator<Gate>::kNormal;
+  if (gate.kind == gate::kMeasurement) {
+    gate_kind = KrausOperator<Gate>::kMeasurement;
+  }
+  return {{gate_kind, 1, 1.0, {gate}}};
+}
+
+void add_gate_channel(const qsim::Cirq::GateKind gate_kind, const unsigned time,
+                      const std::vector<unsigned>& qubits,
+                      const std::map<std::string, float>& params,
+                      NoisyCircuit<Cirq::GateCirq<float>>* ncircuit) {
+  ncircuit->push_back(create_single_gate_channel(
+    create_gate(gate_kind, time, qubits, params)));
+}
+
+void add_diagonal_gate_channel(const unsigned time,
+                               const std::vector<unsigned>& qubits,
+                               const std::vector<float>& angles,
+                               NoisyCircuit<Cirq::GateCirq<float>>* ncircuit) {
+  ncircuit->push_back(create_single_gate_channel(
+    create_diagonal_gate(time, qubits, angles)));
+}
+
+void add_matrix_gate_channel(const unsigned time,
+                             const std::vector<unsigned>& qubits,
+                             const std::vector<float>& matrix,
+                             NoisyCircuit<Cirq::GateCirq<float>>* ncircuit) {
+  ncircuit->push_back(create_single_gate_channel(
+    create_matrix_gate(time, qubits, matrix)));
+}
+
+void control_last_gate_channel(const std::vector<unsigned>& qubits,
+                               const std::vector<unsigned>& values,
+                               NoisyCircuit<Cirq::GateCirq<float>>* ncircuit) {
+  if (ncircuit->back().size() > 1) {
+    throw std::invalid_argument(
+        "Control cannot be added to noisy channels.");
+  }
+  for (Cirq::GateCirq<float>& op : ncircuit->back()[0].ops) {
+    MakeControlledGate(qubits, values, op);
+  }
+}
+
+void add_channel(const unsigned time,
+                 const std::vector<unsigned>& qubits,
+                 const std::vector<std::tuple<float, std::vector<float>, bool>>&
+                     prob_matrix_unitary_triples,
+                 NoisyCircuit<Cirq::GateCirq<float>>* ncircuit) {
+  // Adds a channel to the noisy circuit.
+  using Gate = Cirq::GateCirq<float>;
+  Channel<Gate> channel;
+  // prob_matrix_unitary_triples contains triples with these elements:
+  //   0. The lower-bound probability of applying the matrix.
+  //   1. The matrix to be applied.
+  //   2. Whether the matrix is unitary.
+  for (const auto &triple : prob_matrix_unitary_triples) {
+    const float prob = std::get<0>(triple);
+    const std::vector<float>& mat = std::get<1>(triple);
+    bool is_unitary = std::get<2>(triple);
+    Gate gate = create_matrix_gate(time, qubits, mat);
+    channel.emplace_back(KrausOperator<Gate>{
+      KrausOperator<Gate>::kNormal, is_unitary, prob, {gate}
```

Comment:
Would the need for kMeasurement never come up ?

Reply:
It's not needed in `add_channel`, since measurements don't appear in channels. A measurement gate would instead be handled by `create_single_gate_channel`.